apps/consulting/posts/scaling-python-banking-edition.md

```diff
@@ -0,0 +1,677 @@
+aggregation to each batch. Finally, the workers will re-partition
+and apply the mappings and aggregation across the batches until the
+dataset is fully aggregated.
+
+```python
+def load_dataframe(data):
+    df = pd.read_parquet(
+        data,
+        storage_options=storage_options
+    )
+    df = df.set_index(['Date'], append=True)
+    return df
+
+@delayed
+def info(df):
+    buffer = io.StringIO()
+    df.info(buf=buffer)
+    return buffer.getvalue()
+
+def build_task_graph(filepaths):
+    part_size = 5
+    with dask.annotate(resources={'MEMORY': 200*part_size}):
+        b = db.from_sequence(filepaths, partition_size=part_size)
+        dfs = b.map(load_dataframe)
+        summed_df = dfs.sum()
+    with dask.annotate(resources={'MEMORY': 200}):
+        delayed_df = summed_df.to_delayed(optimize_graph=True)
+    info_delayed = info(delayed_df)
+    return info_delayed
+```
+
+<img
+src="/posts/scaling-python-banking-edition/dask-bag.png"
+alt="A picture of a jupyter notebook input cell, next to an image of the cell contents equivilent dask graph being run. The input cell has the code for a dask bag data frame summation wit resource annotations."
+/>
+
+**_!!!! (We need to walk the reader through the above figure) !!!!_**
+
+On top of Dask Bag, we also started utilizing a tool called 'resource
+annotations.' These allowed us to take a much more
+hands-off approach to the scheduler. Instead of manually managing how much work
+we are submitting at a time, we can submit it all and let the scheduler do the
+heavy lifting of resource management. This automatically insures that the workers will
+not get work that exceeds the resources that they have available.
+
+To use resource annotations, first you need to define the relevant resource
+limits when you set up your cluster.
+
+```python
+with dask.config.set({'distributed.worker.resources.MEMORY': 2000}):
+    cluster = LocalCluster()
+client = Client(cluster)
+```
+
+As you can see in the above example, I used the context manager
+`dask.config.set` to set the `MEMORY` resources of each worker
+to 2000 **_WHAT UNITS?_**. The resource name and the value can be set arbitrarily. For
+example, I could have set the resource to be `WIDGET` instead of
+`MEMORY`, and it would work the same as long as your naming is consistent
+between your cluster configuration and your annotations. Next, you
+annotate your Dask **_Dask Bag?_** objects.
+
+```python
+import dask.bag as db
+
+part_size=10
+with dask.annotate(resources={'MEMORY': 200*part_size}):
+    b = db.from_sequence(filepaths, partition_size=part_size)
+    dfs = b.map(load_dataframe)
+    summed_df = dfs.sum()
+```
+
+In the above code, I have annotated the blocks of Dask objects with
+the max amount of memory I expect them to need. Now, when I run this,
+the scheduler knows not to send off a block of work to a worker until
+there are the necessary resources available on a worker.
```

Comment:
This paragraph also feels like it could use more detail in general. Not having been involved in the project, I can't say for sure, but it seems like there's more that could be explained here.

Reply:
I added a little more context on why this was valuable. Is that what you were looking for?